### PR TITLE
refactor mongodb game reports

### PR DESCRIPTION
**Pull request by TarikViehmann**
This PR fixes https://github.com/robocup-logistics/rcll-refbox/issues/177, https://github.com/robocup-logistics/rcll-refbox/issues/192, https://github.com/robocup-logistics/rcll-refbox/issues/191, https://github.com/robocup-logistics/rcll-refbox/issues/190, https://github.com/robocup-logistics/rcll-refbox/issues/189, https://github.com/robocup-logistics/rcll-refbox/issues/187, https://github.com/robocup-logistics/rcll-refbox/issues/186 and https://github.com/robocup-logistics/rcll-refbox/issues/183.
It also includes and refines the changes from https://github.com/robocup-logistics/rcll-refbox/pull/188, which should be closed in favor of this.

The game report format is thereby significantly altered. The main changes to the format are:
 - snake case on all field names
 - proper gamestate history instead of the prior workaround solution, hence no more `gamestate/<PHASE>`  sub-documents
 - more history information in general, which is also chronologically sorted
 - removal of redundant information `total-points`, `phase-points-magenta` and `phase-points-cyan` 
 - proper storage of types
 
Other than the document format itself, more changes to the surrounding code was necessary. Mainly:
- confval facts now store booleans as proper clips booleans `TRUE`/`FALSE`
- history facts are now maintained independent from the usage of mongodb
- fix misleading error output when loading fields through mongodb caused by the mixed storage of fact templates in mongodb arrays. The fix is  explained further in https://github.com/robocup-logistics/rcll-refbox/commit/a0fd3f02986a0303fef349902fbda8a523f4628e. The error looked like this 
````
14:05:23.487360 C: Loading machine setup from database
14:05:23.579902 C: mongodb-update-fact-from-bson: No rs-meta fact with identifiers (name) found that matches doc values
14:05:23.580325 C: mongodb-update-fact-from-bson: No rs-meta fact with identifiers (name) found that matches doc values
14:05:23.580587 C: mongodb-update-fact-from-bson: No rs-meta fact with identifiers (name) found that matches doc values
14:05:23.580832 C: mongodb-update-fact-from-bson: No rs-meta fact with identifiers (name) found that matches doc values
14:05:23.599757 C: mongodb-update-fact-from-bson: No rs-meta fact with identifiers (name) found that matches doc values
14:05:23.600152 C: mongodb-update-fact-from-bson: No rs-meta fact with identifiers (name) found that matches doc values
14:05:23.600476 C: mongodb-update-fact-from-bson: No rs-meta fact with identifiers (name) found that matches doc values
14:05:23.600755 C: mongodb-update-fact-from-bson: No rs-meta fact with identifiers (name) found that matches doc values
14:05:23.601113 C: mongodb-update-fact-from-bson: No rs-meta fact with identifiers (name) found that matches doc values
14:05:23.601525 C: mongodb-update-fact-from-bson: No rs-meta fact with identifiers (name) found that matches doc values
14:05:23.601803 C: Loading machine-setup finished
14:05:23.611572 C: Loading machine positions from database
````

- properly store an end-time in each report, e.g., when SIGINT is sent at any time. Previously, the end-time was only properly set when the game wen the game went to POST-GAME phase.
- refactoring of game-time to be independent of the game-state, further explained in https://github.com/robocup-logistics/rcll-refbox/commit/6f1f43d68cd4227976912332bb1a607d98e4fb4d
-  globals are now also converted to confval facts and sent to the websocket 
- a bunch of renamings have been taking place to make stuff overall more consistent

This PR is merge ready once:
~~1. The analyze scripts are adapted to the mongodb format changes (@MatteoTschesche can you have a look at that?)~~
~~2. A new grasping challenge field document is generated~~

### CODE REVIEW

**Review by snoato — DISMISSED** [review]
Some minor questions/comments for clarification. Otherwise great revamping. Thanks!

File: src/games/rcll/game.clp [hunk 1]
```diff
@@ -719,3 +740,168 @@
...
+	(create$ ?n-s ?n-ws ?n-hp ?n-mst ?n-mc ?n-ws)
+	(create$ ?s ?ws ?hp ?mst ?mc ?ws))
+	  (and (neq ?n-pose ?pose)
+	       (> (time-diff-sec ?n-pt ?pt) 5))))
+	=>
+	(bind ?rh (modify ?rh (is-latest FALSE)))
+	(duplicate ?rh (state ?n-s) (warning-sent ?n-ws) (has-pose ?n-hp) (pose ?n-pose) (pose-time ?n-pt) (maintenance-start-time ?n-mst) (maintenance-cycles ?n-mc) (maintenance-warning-sent ?n-mws) (time ?now) (game-time ?gt) (is-latest TRUE))
+)
+
+(defrule game-create-first-shelf-slot-history
```
[the remark this answers is not in the document]
Comment (TarikViehmann) [reply src/games/rcll/game.clp:856]: one has as precondition that there is no history fact yet, the other binds the latest history fact and duplicates it.

File: src/games/rcll/mongodb.clp [hunk 2]
```diff
@@ -13,7 +13,7 @@
 	; Mongodb Game Report Version,
 	; 1.1 -> includes config
 	; 1.2 -> mps meta facts, workpiece and agent-task info
-	?*MONGODB-REPORT-VERSION* = 1.2
+	?*MONGODB-REPORT-VERSION* = 2.0
```
[the remark this answers is not in the document]
Comment (TarikViehmann) [reply src/games/rcll/mongodb.clp:16]: in principle, this is a very good idea. Problems are:
1. We have no proper refbox version workflow, despite some tags that we should probably utilize more... The thing is that in the past we encountered and fixed a few things throughout the season.
2. The scope of game reports goes beyond the refbox itself, e.g., other repos such as the mongodb-backend depend on it

What we could do is to associate report versions with seasons/competitions, e.g., 1.2 is for 2023, 2.0 for 2024. The wiki could be a good place to hold that information.

I am also open to different suggestions, what do you think?

File: src/games/rcll/mongodb.clp [hunk 3]
```diff
@@ -413,79 +416,102 @@
...
-		(if ?machine-doc then
-			(bson-append ?history-doc "machine-fact" ?machine-doc)
-		 else
-			(printout warn "mongodb: machine history fact without machine fact!" crlf)
-		)
-		(if ?temp-fact then
-			(retract ?temp-fact)
-		)
+		(bson-append-time ?history-doc "time" (fact-slot-value ?mh time))
+		;(bind ?temp-fact (assert-string ?mh:fact-string))
```
[the remark this answers is not in the document]
Comment (TarikViehmann) [reply src/games/rcll/mongodb.clp]: True, fixed it, thanks!

File: src/games/rcll/mongodb.clp (hunk 2 above)
[the remark this answers is not in the document]
Comment (snoato) [reply src/games/rcll/mongodb.clp:16]: I think something as simple as version = season and keeping that in the wiki would be good enough!

**Review by snoato — APPROVED** [review]
Great work!

**Pull request merged**